src/WebJobs.Script.WebHost/StartupContextProvider.cs

```diff
@@ -0,0 +1,149 @@
+        private StartupContext _startupContext;
+
+        public StartupContextProvider(IEnvironment environment, ILogger<StartupContextProvider> logger)
+        {
+            _environment = environment;
+            _logger = logger;
+        }
+
+        internal StartupContext Context
+        {
+            get
+            {
+                // context is only loaded once on startup
+                if (!_loaded)
+                {
+                    _startupContext = GetStartupContextOrNull();
+                    _loaded = true;
+                }
+                return _startupContext;
+            }
+
+            set
+            {
+                _startupContext = value;
+                _loaded = true;
+            }
+        }
+
+        public virtual HostSecretsInfo GetHostSecretsOrNull()
+        {
+            if (Context?.Secrets?.Host != null)
+            {
+                var hostSecrets = Context.Secrets.Host;
+                var secretsInfo = new HostSecretsInfo
+                {
+                    MasterKey = hostSecrets.Master,
+                    FunctionKeys = new Dictionary<string, string>(hostSecrets.Function, StringComparer.OrdinalIgnoreCase),
+                    SystemKeys = new Dictionary<string, string>(hostSecrets.System, StringComparer.OrdinalIgnoreCase)
+                };
+                _logger.LogDebug("Loaded host keys from startup context");
+
+                return secretsInfo;
+            }
+
+            return null;
+        }
+
+        public virtual IDictionary<string, IDictionary<string, string>> GetFunctionSecretsOrNull()
+        {
+            if (Context?.Secrets?.Function != null)
+            {
+                var functionKeys = Context.Secrets.Function.ToDictionary(p => p.Name, p => p.Secrets);
+
+                _logger.LogDebug($"Loaded keys for {functionKeys.Keys.Count} functions from startup context");
+
+                return functionKeys;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Load context from local file system if specified.
+        /// </summary>
+        /// <returns>The context.</returns>
+        private StartupContext GetStartupContextOrNull()
+        {
+            var contextPath = _environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebsiteStartupContextCache);
+            if (!string.IsNullOrEmpty(contextPath))
+            {
+                try
+                {
+                    _logger.LogDebug($"Loading startup context from {contextPath}");
+                    string content = File.ReadAllText(contextPath);
+
+                    // Context files are onetime use. We delete after reading to ensure
+                    // that we don't use a stale file in the future if the app recycles, etc.
+                    // Dont' want to block on this file operation, so we kick it off in the background.
+                    Task.Run(() => File.Delete(contextPath));
```

Comment:
No its not shared across all workers - it's a local file. If the app recycles, no we don't wan't to reload it - we want it to be a one time read.  DWAS won't be refreshing it so it could become stale over time, so we only want to use it the first time we get it on specialization. Correct - it's coming from the local file system, not shared files. If the delete fails it's not the end of the world which is why I didn't log. The cache invalidation I have elsewhere will handle it if the cache is stale for whatever reason.